### PR TITLE
Fat trait fix attempt 2

### PR DESCRIPTION
**Pull request by Varoxus**
## About The Pull Request
Adds in a new trait to handle the tackle defense bonus that was previously granted by `TRAIT_FAT` to solve a few problems that were occurring downstream, this new trait is handled by the stomach organ and handles it all automatically. Also edits the lipoplasty check to see that the target is fat from eating too much.

## Why It's Good For The Game
No functions will have their results altered by the change and it will help solve some some issues downstream.

### CODE REVIEW

**Comment by MrMelbert** [comment]
I understand why it's being done but I also don't know what is the point of `TRAIT_FAT` if it only ever has side effects if added by a single source.

**Comment by Varoxus** [comment]
> I understand why it's being done but I also don't know what is the point of `TRAIT_FAT` if it only ever has side effects if added by a single source.

The point is `TRAIT_FAT` applies a lot of different effects when a mob is assigned it. It doesn't matter if it's only from one source since the one source you can get it from is the only way to currently get it. As stated, this isn't much of a problem for the base TG codebase itself, but this solution also won't interfere with anything and will solve problems downstream where it's more of an issue.

**Comment by Varoxus** [comment]
and if the concern is "What if someone wanted to add another source to get the trait", all they would need to do is modify the current `if(HAS_TRAIT_FROM(mob, TRAIT_FAT, OBESITY))` checks into `if(HAS_TRAIT_FROM(mob, TRAIT_FAT, OBESITY)) (&& or ||) if(HAS_TRAIT_FROM(mob, TRAIT_FAT, NEW_TRAIT))`.

**Comment by MrMelbert** [comment]
> and if the concern is "What if someone wanted to add another source to get the trait", all they would need to do is modify the current `if(HAS_TRAIT_FROM(mob, TRAIT_FAT, OBESITY))` checks into `if(HAS_TRAIT_FROM(mob, TRAIT_FAT, OBESITY) (&& or ||) HAS_TRAIT_FROM(mob, TRAIT_FAT, NEW_TRAIT))`.

This is... bad tho. Someone shouldn't have to go through and update every single consumer of the trait if they want to add it via a different source.

**Comment by MrMelbert** [comment]
> The point is `TRAIT_FAT` applies a lot of different effects when a mob is assigned it. It doesn't matter if it's only from one source since the one source you can get it from is the only way to currently get it. As stated, this isn't much of a problem for the base TG codebase itself, but this solution also won't interfere with anything and will solve problems downstream where it's more of an issue.

It'd probably be easier if you just split off the effects you *don't* want to be affected by `TRAIT_FAT` into a different trait, and then to use that trait instead.

File: code/modules/surgery/lipoplasty.dm [hunk 1]
```diff
@@ -10,7 +10,7 @@
 	)
 
 /datum/surgery/lipoplasty/can_start(mob/user, mob/living/carbon/target)
-	if(!HAS_TRAIT(target, TRAIT_FAT) || target.nutrition < NUTRITION_LEVEL_WELL_FED)
+	if(!HAS_TRAIT_FROM(target, TRAIT_FAT, OBESITY) || target.nutrition < NUTRITION_LEVEL_WELL_FED)
```
Comment (MrMelbert) [inline code/modules/surgery/lipoplasty.dm:13]: This one is probably fine, though.

**Comment by Varoxus** [comment]
> > The point is `TRAIT_FAT` applies a lot of different effects when a mob is assigned it. It doesn't matter if it's only from one source since the one source you can get it from is the only way to currently get it. As stated, this isn't much of a problem for the base TG codebase itself, but this solution also won't interfere with anything and will solve problems downstream where it's more of an issue.
> 
> It'd probably be easier if you just split off the effects you _don't_ want to be affected by `TRAIT_FAT` into a different trait, and then to use that trait instead.
> 
> Or visa versa: split off ONLY the effects you want affected by `TRAIT_FAT` into a new trait. From what I see the only one missing is the tackle bonus? So change that into a new trait.

Could do, my only concern would be that I'd be defining a new trait that's only used in one check (the tackle), and that trait wouldn't have any method from which a mob could actually gain it. Unless I'm not not understanding what you mean, exactly.

**Comment by MrMelbert** [comment]
> Could do, my only concern would be that I'd be defining a new trait that's only used in one check (the tackle), and that trait wouldn't have any method from which a mob could actually gain it. Unless I'm not not understanding what you mean, exactly. Honestly though, if that were an acceptable solution to the maintainers, then I'd be more than fine simply doing that instead.

Nothing wrong with that, could make it generic like `TRAIT_OFF_BALANCE_TACKLER` which is just a simple tackle malus and that's that

**Comment by Varoxus** [comment]
Alright, trying that out. If it's accepted I can update the description and such

File: code/datums/components/tackle.dm [hunk 2]
```diff
@@ -369,7 +369,7 @@
 
 	if(HAS_TRAIT(target, TRAIT_CLUMSY))
 		defense_mod -= 2
-	if(HAS_TRAIT(target, TRAIT_FAT)) // chonkers are harder to knock over
+	if(HAS_TRAIT(target, TRAIT_FAT) || HAS_TRAIT(target, TRAIT_OFF_BALANCE_TACKLER)) // chonkers are harder to knock over
```
Comment (MrMelbert) [inline code/datums/components/tackle.dm:372]: Alright now you would remove the `HAS_TRAIT(target, TRAIT_FAT)` and instead, when we `ADD_TRAIT(mob, TRAIT_FAT, OBESITY` you would have `ADD_TRAIT(mob, TRAIT_OFF_BALANCE_TACKLER, OBESITY)`

**Comment by Ghommie** [comment]
if would be nice if the changelog were written in a way that was understandable by the average player that doesn't look at the code.

**Comment by Varoxus** [comment]
> if would be nice if the changelog were written in a way that was understandable by the average player that doesn't look at the code.

tried my best, but considering this is basically all code stuff  and won't affect the average player... dunno if there's really a good way to put it. I'd take any suggestions though.

**Comment by san7890** [comment]
> tried my best, but considering this is basically all code stuff and won't affect the average player... dunno if there's really a good way to put it. I'd take any suggestions though.

The second entry is probably the only thing that could be considered useful out of the three, but me myself personally I would just omit all three since it doesn't affect players.

File: code/modules/mob/living/carbon/human/init_signals.dm [hunk 3]
```diff
@@ -60,6 +60,7 @@
 		satiety = 0
 		overeatduration = 0
 		REMOVE_TRAIT(src, TRAIT_FAT, OBESITY)
+		REMOVE_TRAIT(src, TRAIT_OFF_BALANCE_TACKLER, OBESITY)
```
Comment (san7890) [inline code/modules/mob/living/carbon/human/init_signals.dm:63]: you can (and should) use `remove_traits()` here

```suggestion
		remove_traits(list(TRAIT_FAT, TRAIT_OFF_BALANCE_TACKLER), OBESITY)
```

File: code/modules/surgery/organs/internal/stomach/_stomach.dm [hunk 4]
```diff
@@ -128,11 +128,13 @@
 		if(human.overeatduration < (200 SECONDS))
 			to_chat(human, span_notice("You feel fit again!"))
 			REMOVE_TRAIT(human, TRAIT_FAT, OBESITY)
+			REMOVE_TRAIT(human, TRAIT_OFF_BALANCE_TACKLER, OBESITY)
```
Comment (san7890) [inline code/modules/surgery/organs/internal/stomach/_stomach.dm:131]:
```suggestion
			human.remove_traits(list(TRAIT_FAT, TRAIT_OFF_BALANCE_TACKLER), OBESITY)
```

File: code/modules/surgery/organs/internal/stomach/_stomach.dm [hunk 5]
```diff
@@ -128,11 +128,13 @@
...
 
 	else
 		if(human.overeatduration >= (200 SECONDS))
 			to_chat(human, span_danger("You suddenly feel blubbery!"))
 			ADD_TRAIT(human, TRAIT_FAT, OBESITY)
+			ADD_TRAIT(human, TRAIT_OFF_BALANCE_TACKLER, OBESITY)
```
Comment (san7890) [inline code/modules/surgery/organs/internal/stomach/_stomach.dm:137]:
```suggestion
			human.add_traits(list(TRAIT_FAT, TRAIT_OFF_BALANCE_TACKLER), OBESITY)
```

**Review by san7890 — CHANGES_REQUESTED** [review]
prefer to use `add_traits()` and `remove_traits()` in places like these to make it a bit more succinct

**Comment by Varoxus** [comment]
Alrighty, code tidied up and CL removed since it won't affect players.

**Pull request merged**